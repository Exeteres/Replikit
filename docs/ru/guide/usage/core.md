# Ядро

`@replikit/core` - модуль, без которого не будет работать ни одно Replikit-приложение.
Он отвечает за запуск контроллеров, предоставляет системы конфигурации и хуков.
Все интерфейсы, используемые или реализуемые контроллерами также объявлены в этом модуле.

## События

Одна из главных задач ядра - обработка событий, получаемых от контроллеров.
Для этого используется один единственный обработчик, который задается функцией `setHandler`.
Как правило, этот обработчик автоматически устанавливается и используется пакетом `@replikit/router`.

## Конфигурация

В ядре также располагается централизованная система конфигурации для различных модулей.
Для её обновления существует функция `updateConfig`, которая рекурсивно применит изменения из полученного объекта и вызовет хук `core:settings:update`.

```ts
import { updateConfig, config } from "@replikit/core";

updateConfig({
    mymodule: {
        field: "value"
    }
});

console.log(config.mymodule.field);
```

## Логирование

Логирование осуществляется экземплярами логгера, каждый из которых может иметь свою область видимости.

```ts
import { createScope } from "@replikit/core";

// Создаем область видимости mymodule
const logger = createScope("mymodule");
logger.info("Hi!");
```

Результат:

```
[INF] [XXXX-XX-XX GMTXXX:XX XX:XX:XX] [mymodule] Hi!
```

Области видимости также могут быть вложенными:

```ts
import { createScope } from "@replikit/core";

const logger = createScope("mymodule");

// Создаем область видимости submodule внтури mymodule
const nestedLogger = logger.createScope("submodule");
nestedLogger.info("Hi from submodule!");
```

Результат:

```
[INF] [XXXX-XX-XX GMTXXX:XX XX:XX:XX] [mymodule] [submodule] Hi from submodule!
```

## Хуки

Система хуков позволяет модулям выполнять код на различных этапах как загрузки ядра, так и его использования.
Обработчики хуков могут быть асинхронными. Они выполняются независимо друг от друга.

```ts
import { hook } from "@replikit/core";

hook("core:startup:init", () => {
    // Будет вызван перед загрузкой ядра
    console.log("init")
});

hook("core:startup:ready", () => {
    // Будет вызван когда ядро готово обрабатывать сообщения
    console.log("ready")
});

```

Для вызова хуков существует метод `invokeHook`, однако вы никогда не должны самостоятельно вызывать хуки сторонних модулей.
Это может нарушить работу всей системы.

Хуки ядра:

| Хук                    | Описание                                                                                  |
| ---------------------- | ----------------------------------------------------------------------------------------- |
| `core:startup:init`    | Начало процесса запуска                                                                   |
| `core:startup:done`    | Запуск завершился, но контроллеры еще не запущены                                         |
| `core:startup:ready`   | Ядро готово обрабатывать сообщения                                                        |
| `core:shutdown:init`   | Начало процесса завершения                                                                |
| `core:shutdown:done`   | Вызывается перед завершением процесса                                                     |
| `core:settings:update` | Конфигурация обновилась; обработчики получают копию старой конфигурации первым аргументом |

Хуки других модулей:

| Хук                      | Описание                           |
| ------------------------ | ---------------------------------- |
| `storage:database:init`  | Вызывается перед подключением к бд |
| `storage:database:done`  | Вызывается после подключения       |
| `storage:database:ready` | Хранилище готово к использованию   |
