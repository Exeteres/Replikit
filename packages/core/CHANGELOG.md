# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [0.17.2](https://github.com/Exeteres/Replikit/compare/v0.17.1...v0.17.2) (2020-08-24)

**Note:** Version bump only for package @replikit/core





## [0.17.1](https://github.com/Exeteres/Replikit/compare/v0.17.0...v0.17.1) (2020-08-21)

**Note:** Version bump only for package @replikit/core





# [0.17.0](https://github.com/Exeteres/Replikit/compare/v0.16.0...v0.17.0) (2020-08-21)


### Features

* **core:** Add Identifier type ([e97ef90](https://github.com/Exeteres/Replikit/commit/e97ef90d24068fe4c16842b17bae0b8e6570ec47))





# [0.16.0](https://github.com/Exeteres/Replikit/compare/v0.15.1...v0.16.0) (2020-08-18)


### Bug Fixes

* **core:** Fix bug in text tokenization system ([43404bf](https://github.com/Exeteres/Replikit/commit/43404bffdb878e1f4abd18f69970cd6520251dd9))


### Features

* **core:** Add support for message headers ([5754c94](https://github.com/Exeteres/Replikit/commit/5754c94e0f89edb4ff3dde4fe410ade303bfb58f))





## [0.15.1](https://github.com/Exeteres/Replikit/compare/v0.15.0...v0.15.1) (2020-08-13)

**Note:** Version bump only for package @replikit/core





# [0.15.0](https://github.com/Exeteres/Replikit/compare/v0.14.0...v0.15.0) (2020-08-13)

**Note:** Version bump only for package @replikit/core





# [0.14.0](https://github.com/Exeteres/Replikit/compare/v0.13.0...v0.14.0) (2020-08-05)


### Bug Fixes

* **cli:** Fix typings ([9851fe5](https://github.com/Exeteres/Replikit/commit/9851fe5b87e07318da5c085f4b1bbddb23984136))





# [0.13.0](https://github.com/Exeteres/Replikit/compare/v0.12.1...v0.13.0) (2020-08-04)


### Bug Fixes

* **core:** Fix constructor type ([9687c48](https://github.com/Exeteres/Replikit/commit/9687c48773917919a2d8c38add27fba15cc11f00))


### Features

* **core:** Add ability to use pipe method in builders ([95c3cca](https://github.com/Exeteres/Replikit/commit/95c3cca0d1f4c5203fde3386c3775ecc77b92a1b))





# [0.12.0](https://github.com/Exeteres/Replikit/compare/v0.11.2...v0.12.0) (2020-07-30)


### Bug Fixes

* **core:** Rename query field to text ([4b0a930](https://github.com/Exeteres/Replikit/commit/4b0a930a4a840efcf56bdc6dfcf0b77a9430a30c))


### Features

* **core:** Refactor types and add support for inline queries ([2e52258](https://github.com/Exeteres/Replikit/commit/2e52258587cf0f2d31061c991eb2489ec7acda9e))





## [0.11.2](https://github.com/Exeteres/Replikit/compare/v0.11.1...v0.11.2) (2020-07-24)

**Note:** Version bump only for package @replikit/core





## [0.11.1](https://github.com/Exeteres/Replikit/compare/v0.11.0...v0.11.1) (2020-07-24)

**Note:** Version bump only for package @replikit/core






# [0.11.0](https://github.com/Exeteres/Replikit/compare/v0.10.1...v0.11.0) (2020-07-24)


### Bug Fixes

* **core:** Fix deepmerge ([132b184](https://github.com/Exeteres/Replikit/commit/132b18493e02603c195af70a2cfff0e9c31dc383))


### Features

* **sessions:** Add sessions package ([e7a0adf](https://github.com/Exeteres/Replikit/commit/e7a0adf3bbe35091bad6dcf28395e3c389a20a83))





## [0.10.1](https://github.com/Exeteres/Replikit/compare/v0.10.0...v0.10.1) (2020-06-30)

**Note:** Version bump only for package @replikit/core






# [0.10.0](https://github.com/Exeteres/Replikit/compare/v0.9.0...v0.10.0) (2020-06-29)

**Note:** Version bump only for package @replikit/core





# [0.9.0](https://github.com/Exeteres/Replikit/compare/v0.8.0...v0.9.0) (2020-06-28)


### Features

* Add advanced metadata processing ([54c7820](https://github.com/Exeteres/Replikit/commit/54c782021f5ebc16d784f210d273f8d7490d2af5))





# [0.8.0](https://github.com/Exeteres/Replikit/compare/v0.7.2...v0.8.0) (2020-06-27)


### Features

* **core:** Add core support for replies ([21f3318](https://github.com/Exeteres/Replikit/commit/21f3318f7bd031233856e7a1e1ddd29d73ef5f04))





## [0.7.2](https://github.com/Exeteres/Replikit/compare/v0.7.1...v0.7.2) (2020-06-23)

**Note:** Version bump only for package @replikit/core





# [0.7.0](https://github.com/Exeteres/Replikit/compare/v0.6.0...v0.7.0) (2020-06-23)

**Note:** Version bump only for package @replikit/core





# [0.6.0](https://github.com/Exeteres/Replikit/compare/v0.5.3...v0.6.0) (2020-06-21)

**Note:** Version bump only for package @replikit/core





## [0.5.3](https://github.com/Exeteres/Replikit/compare/v0.5.2...v0.5.3) (2020-06-20)

**Note:** Version bump only for package @replikit/core





## [0.5.2](https://github.com/Exeteres/Replikit/compare/v0.5.1...v0.5.2) (2020-06-20)

**Note:** Version bump only for package @replikit/core





# 0.5.0 (2020-06-20)


### Features

* **cli:** Add build command ([5c9dc57](https://github.com/Exeteres/Replikit/commit/5c9dc57b9ac2428f4970096dcfc6ae75b2e4dd27))






# [0.4.0](https://github.com/Exeteres/Replikit/compare/v0.3.0...v0.4.0) (2020-06-19)


### Features

* **cli:** Add build command ([0212d7f](https://github.com/Exeteres/Replikit/commit/0212d7f9c87d517b8cf220be9b32376551c3603b))





# [0.3.0](https://github.com/Exeteres/Replikit/compare/v0.2.0...v0.3.0) (2020-06-14)

**Note:** Version bump only for package @replikit/core





# [0.2.0](https://github.com/Exeteres/Replikit/compare/v0.1.0...v0.2.0) (2020-06-13)

**Note:** Version bump only for package @replikit/core
