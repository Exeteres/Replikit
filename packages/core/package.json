{
    "name": "@replikit/core",
    "version": "0.11.1",
    "license": "MIT",
    "main": "dist/index.js",
    "typings": "dist/index.d.ts",
    "files": [
        "dist",
        "typings"
    ],
    "scripts": {
        "build": "tsc -p tsconfig.build.json"
    }
}
