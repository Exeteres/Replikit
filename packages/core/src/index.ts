/// <reference path="../typings/index.d.ts" />

export * from "./models/channelType";
export * from "./models/attachmentType";
export * from "./models/textTokenKind";
export * from "./models/textTokenProp";
export * from "./models/logLevel";

export * from "./composition/compositionItem";
export * from "./composition/compositionFactory";
export * from "./composition/compositionContext";
export * from "./composition/composition";

export * from "./errors";
export * from "./utils";
export * from "./hookStorage";
export * from "./config";
export * from "./eventHandler";
export * from "./cacheManager";
export * from "./textTokenizer";
export * from "./textFormatter";
export * from "./controller";
export * from "./controllerStorage";

export * from "./logging/consoleLogProvider";
export * from "./logging/logManager";
export * from "./logging/logger";

import "./startup";
