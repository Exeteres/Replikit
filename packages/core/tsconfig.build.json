{
    "extends": "../../tsconfig.build.json",
    "compilerOptions": {
        "outDir": "dist",
        "paths": {
            "@replikit/core": ["core/src"]
        }
    },
    "include": ["src"]
}
