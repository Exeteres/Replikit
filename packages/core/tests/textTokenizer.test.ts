import { TextTokenizer, TextTokenKind, TextTokenProp } from "@replikit/core";

const tokenizer = new TextTokenizer()
    .addTextPropRule("**", TextTokenProp.Bold)
    .addTextPropRule("*", TextTokenProp.Italic)
    .addTextPropRule("__", TextTokenProp.Underline)
    .addTextPropRule("~~", TextTokenProp.Strikethrough)
    .addTextPropRule("```", TextTokenProp.Monospace)
    .addRegexRule(/<!?@(\d*)>/, groups => ({
        kind: TextTokenKind.Mention as const,
        id: +groups[1],
        props: []
    }));

describe("textTokenizer", () => {
    it("should tokenize a text with props", () => {
        const text = "**test** __test__ *test*";
        const tokens = tokenizer.tokenize(text);

        expect(tokens).toMatchSnapshot();
    });

    it("should tokenize a text with mentions and props", () => {
        const text = "<!@88005553535> __***test*** ~~test~~__ <@999>";
        const tokens = tokenizer.tokenize(text);

        expect(tokens).toMatchSnapshot();
    });

    it("should tokenize a text with multiline code", () => {
        const text = "test ```\ntest\n```";
        const tokens = tokenizer.tokenize(text);

        expect(tokens).toMatchSnapshot();
    });
});
