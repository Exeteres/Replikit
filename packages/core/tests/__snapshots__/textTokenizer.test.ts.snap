// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`textTokenizer should tokenize a text with mentions and props 1`] = `
Array [
  Object {
    "id": 88005553535,
    "kind": 1,
    "props": Array [],
  },
  Object {
    "kind": 0,
    "props": Array [],
    "text": " ",
  },
  Object {
    "kind": 0,
    "props": Array [
      3,
      1,
      2,
    ],
    "text": "test",
  },
  Object {
    "kind": 0,
    "props": Array [
      3,
    ],
    "text": " ",
  },
  Object {
    "kind": 0,
    "props": Array [
      3,
      4,
    ],
    "text": "test",
  },
  Object {
    "kind": 0,
    "props": Array [],
    "text": " ",
  },
  Object {
    "id": 999,
    "kind": 1,
    "props": Array [],
  },
]
`;

exports[`textTokenizer should tokenize a text with multiline code 1`] = `
Array [
  Object {
    "kind": 0,
    "props": Array [],
    "text": "test ",
  },
  Object {
    "kind": 0,
    "props": Array [
      5,
    ],
    "text": "
test
",
  },
]
`;

exports[`textTokenizer should tokenize a text with props 1`] = `
Array [
  Object {
    "kind": 0,
    "props": Array [
      1,
    ],
    "text": "test",
  },
  Object {
    "kind": 0,
    "props": Array [],
    "text": " ",
  },
  Object {
    "kind": 0,
    "props": Array [
      3,
    ],
    "text": "test",
  },
  Object {
    "kind": 0,
    "props": Array [],
    "text": " ",
  },
  Object {
    "kind": 0,
    "props": Array [
      2,
    ],
    "text": "test",
  },
]
`;
